ratelimiter/rsyslogstyle.go

```diff
@@ -0,0 +1,78 @@
+package ratelimiter
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/segmentio/rate-limiting-log-proxy/logger"
+)
+
+// RsyslogStyle is a rate limiter that works similar to rsyslog (and journald's)
+// rate limiters, allowing you to set an interval and a burst.  If there are more
+// messages than burst in a given interval, all further messages are dropped
+// until the end of that interval.
+type RsyslogStyle struct {
+	Interval time.Duration
+	Burst    int
+
+	start  time.Time
+	count  int
+	missed int
+	mu     sync.Mutex
+}
+
+// NewRsyslogStyle returns a RateLimiter based on rsyslog's rate limiter
+func NewRsyslogStyle(interval time.Duration, burst int) *RsyslogStyle {
+	return &RsyslogStyle{
+		Interval: interval,
+		Burst:    burst,
+	}
+}
+
+// Limit should be called each time you want to test if you've reached your rate
+// limit.  It will return true if you are over the limit and false otherwise.
+func (r *RsyslogStyle) Limit(logger logger.Logger) bool {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	now := time.Now()
+
+	if r.start.IsZero() {
+		r.start = now
+	}
+
+	if now.After(r.start.Add(r.Interval)) {
+		if r.missed > 0 {
+			// TODO: send datadog stats about dropped messages
+			logger.Log(fmt.Sprintf("dropped %d messages", r.missed))
```

Comment:
Any thoughts on ensuring that this gets printed (in the case of a shutdown or a beginning burst of logs if no more are generated)? I'm not sure if we want to have the perf hit of having a ticker, but wanted to raise the issue.